swarm/storage/ldbstore.go

```diff
@@ -523,43 +568,54 @@ func (s *LDBStore) CurrentStorageIndex() uint64 {
 	return s.dataIdx
 }
 
-func (s *LDBStore) Put(ctx context.Context, chunk *Chunk) {
+func (s *LDBStore) Put(ctx context.Context, chunk Chunk) error {
 	metrics.GetOrRegisterCounter("ldbstore.put", nil).Inc(1)
-	log.Trace("ldbstore.put", "key", chunk.Addr)
+	log.Trace("ldbstore.put", "key", chunk.Address())
 
-	ikey := getIndexKey(chunk.Addr)
+	ikey := getIndexKey(chunk.Address())
 	var index dpaDBIndex
 
-	po := s.po(chunk.Addr)
+	po := s.po(chunk.Address())
+
 	s.lock.Lock()
-	defer s.lock.Unlock()
 
-	log.Trace("ldbstore.put: s.db.Get", "key", chunk.Addr, "ikey", fmt.Sprintf("%x", ikey))
+	if s.closed {
+		s.lock.Unlock()
+		return ErrDBClosed
+	}
+	batch := s.batch
+
+	log.Trace("ldbstore.put: s.db.Get", "key", chunk.Address(), "ikey", fmt.Sprintf("%x", ikey))
 	idata, err := s.db.Get(ikey)
 	if err != nil {
 		s.doPut(chunk, &index, po)
-		batchC := s.batchC
-		go func() {
-			<-batchC
-			chunk.markAsStored()
-		}()
 	} else {
-		log.Trace("ldbstore.put: chunk already exists, only update access", "key", chunk.Addr)
+		log.Trace("ldbstore.put: chunk already exists, only update access", "key", chunk.Address)
 		decodeIndex(idata, &index)
-		chunk.markAsStored()
 	}
 	index.Access = s.accessCnt
 	s.accessCnt++
 	idata = encodeIndex(&index)
 	s.batch.Put(ikey, idata)
+
+	s.lock.Unlock()
+
 	select {
 	case s.batchesC <- struct{}{}:
 	default:
 	}
+
+	select {
+	case <-batch.c:
+		return batch.err
+	case <-ctx.Done():
+		// log.Error("context done", "err", ctx.Err())
```

Comment:
Remove comment or enable?

Reply:
Fixed in 9f7031a7cd0c555cf0bbc8232709aed5423a464e